### DISCUSSION TITLE
What's the purpose of question-mark rule in dev?

### DISCUSSION THREAD

**Discussion by jods4** [unocss/unocss#3499] [General]
(I'm using Wind preset).
After including JS files in UnoCss, I noticed a lot of files triggered the inclusion of a weird `.?` class.

First surprise: it seems like UnoCss matches any text in a JS file, not just inside strings. That seems like a missed opportunity:
- for perf as anything that's outside strings does not need to be tokenized and matched;
- it creates many false positive: classes that are commonly found in code but not actually used, such as `?` but also `static` keyword, and other variable names.

I found that this weird class comes from here and is only included in `dev` mode:
https://github.com/unocss/unocss/blob/main/packages/preset-mini/src/_rules/question-mark.ts

What's its purpose?

**Comment by userquin** [comment]
Check the footer in [this playground](https://unocss.dev/play/?html=DwEwlgbgBAFgtAMwK4BsVQC4FMAeG4DGWAdtgE5QIq5QDOW1B%2BxA9sVlAIZoBcGZnYrTAYwbOABYADFIB8AKChRQkKAFtOCpUpXRseOAFYc6BAHcAjDK7EwG7HABGLJMSJxuGG3c4OCL0jgwYgRgkQ5BHwcQJAFRcQtaLW0lAFVWAGEAZSzFbWAAenAIZJ1iqBYABwBmKUxcfBQAc0ozWrq1C1LtABUYDmDaDEEvcRAsDWIQKABBDBY1MAIobKyoEibgrAA6PJ0iyG7ddQAmSmocKAArJCGwBABPQhJyeoMTkwqaupgWCCwyABeABEVQAHFJgVwUGAmsQ4BYAKTdHScPYpKBLcQoFhNFhwTYYGBIRzolL6fDBfpkERk7QwMhYBAgmAYDCVWg8AoFQnExzbfxqAquFgEWi0YWsMW0YF0pTDMhNLAYEEAfUcKEEAGtZRioLJCpp0YViskTYd5OaSpbymZqRFHLQWCgkNgoM42QsjFAaU1WXA6tQEPg6hTnqQAV9aq12skxhNBNMAO2YoTDUhQFPINAPKAEW7zRYAL04GqwloOJSAA&config=JYWwDg9gTgLgBAbwFBzgEwKYDNgDsMDCEuOA5gDQpxhQYDOGMAgjDFMAEYCuMwWAnpVQ16jAJIBjYnSHVaDGAFVcESgF84WKBBBwA5FxUS6dPUiQYAHpFjpsAQy4AbeJhz4iJYKQAUyVBi4AG4AshCYAFz6mEF6slDO9FEA2lSoyXoSXHQwOgC0CU4YcYhwUk7QUXq0aHpwagC6VA2ydAAW0DBZMHQpaaWZ2bkgee2d3XpVMFYweU6kcNOWs9D2uKQYcB1BGFARS7PT9k51arLpAPQAesBSuHk%2BAHQAVACUACQX5HA%2Byd-ADVecAAvAA%2BOAAA2AeQk9igHGIEXeCGAaghTVQLSoIgUvTgqVQwnkjGUEB8r3OclEzFY7G4vAE5MpOPEdzofn6qDosKKUQAjI8AEyU1ASNC4KptVhgXoXC70ECPdoXMyE%2BoU5pINSvJBAA&css=PQKgBA6gTglgLgUzAYwK4Gc4HsC2YDCAyoWABYJQIA0YAhgHYAmYcUD6AZllDhWOqgAOg7nAB0YAGLcwCAB60cggDYIAXGBDAAUKDBi0mXGADe2sGC704AWgDuCGAHNScDQFYADJ4Dc5sAACtMLKAJ5gggCMLPK2ABR2pPBIcsoAlH4WAEa0yADWTlBYqEw2yFjK3Bpw5LxxAOTllVDoYpSMYgAs3vUZ2gC%2BmsBAA&options=N4XyA) (using where in the footer or change it to `class="?"`)

EDIT: must be explicitly enabled with `envMode: 'dev'`, check the configuration

**Reply by jods4** [reply]
I did NOT explicitly enable it, at least when using Vite `envMode` defaults to `dev` during development. 
Quick search in code source finds at least 2 places where this might happen: 
https://github.com/unocss/unocss/blob/901db3459b3eff16e230a820549038f28518a722/packages/vite/src/index.ts#L36
and
https://github.com/unocss/unocss/blob/901db3459b3eff16e230a820549038f28518a722/packages/vite/src/config-hmr.ts#L11-L12

I see the footer blinking but what's the purpose? I'll extrapolate from the `where` alias of `?`: is it a tool when you look for an element on the page and you have no idea where it is? (there are other ways... but why not, assuming it's even visible).

I'm fine with the idea, but it's somewhat unsettling if you don't know nor expect it.

I first noticed a weird inclusion of a class `\.?` after enabling JS parsing. 
To me, that looked like a super weird class name that shouldn't be in `presetWind()`.
I looked in the UnoCSS docs, found nothing.
I thought that might be a Regex bug, where one rules matches `?` unexpectedly and did some research.
I found the actual rule in source code and noticed it's not a bug but by design, without much infos.

That's some time lost on this.
It would be cool to maybe generate a comment in the rule output (`/* This is a helper to highlight elements in DEV mode only */`), add it to the documentation, and maybe give up `?`, as `where` is more descriptive and does the job just as well.
-- or make this a configurable opt-in, maybe more discoverable.

Thanks for the playground!

**Reply by userquin** [reply]
It is a porting of windicss plugin https://windicss.org/plugins/community/question-mark.html

We need to add some entry to the docs.

**Reply by jods4** [reply]
Interesting, thanks!
I'm a Tailwind user myself, which is why this was unfamiliar to me.
The class name makes sense now, yeah maybe it just needs to be added to the docs.
